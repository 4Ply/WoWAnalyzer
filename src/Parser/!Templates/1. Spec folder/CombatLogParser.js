import CoreCombatLogParser from 'Parser/Core/CombatLogParser';

// TODO: Pick the ones relevant to your spec:
// import DamageDone from 'Parser/Core/Modules/DamageDone';
// import HealingDone from 'Parser/Core/Modules/HealingDone';
// import DamageTaken from 'Parser/Core/Modules/DamageTaken';

import Abilities from './Modules/Abilities';
import Checklist from './Modules/Features/Checklist';
import AlwaysBeCasting from './Modules/Features/AlwaysBeCasting';

// TODO: import CooldownThroughputTracker from './Modules/Features/CooldownThroughputTracker';

class CombatLogParser extends CoreCombatLogParser {
  static specModules = {
    // TODO: Enable the ones relevant to your spec:
    // damageDone: [DamageDone, { showStatistic: true }],
    // healingDone: [HealingDone, { showStatistic: true }],
    // damageTaken: [DamageTaken, { showStatistic: true }],

    abilities: Abilities, // TODO: Go to Abilities class and fill it with all abilities available to your spec

    // Features
    checklist: Checklist, // TODO: Go into Checklist to define rules
    alwaysBeCasting: AlwaysBeCasting, // TODO: Define a list of all abilities on the GCD
    // TODO: cooldownThroughputTracker: CooldownThroughputTracker,
  };
}

export default CombatLogParser;
