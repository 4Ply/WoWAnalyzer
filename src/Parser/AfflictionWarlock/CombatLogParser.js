import React from 'react';

import SuggestionsTab from 'Main/SuggestionsTab';
import Tab from 'Main/Tab';
import Talents from 'Main/Talents';

import MainCombatLogParser from 'Parser/Core/CombatLogParser';

import CastEfficiency from './Modules/Features/CastEfficiency';
import AlwaysBeCasting from './Modules/Features/AlwaysBeCasting';
import CooldownTracker from './Modules/Features/CooldownTracker';
import AgonyUptime from './Modules/Features/AgonyUptime';
import CorruptionUptime from "./Modules/Features/CorruptionUptime";
import SiphonLifeUptime from "./Modules/Features/SiphonLifeUptime";

import SoulShardTracker from "./Modules/SoulShards/SoulShardTracker";
import SoulShardDetails from "./Modules/SoulShards/SoulShardDetails";

class CombatLogParser extends MainCombatLogParser {

  static specModules = {
    // Features
    castEfficiency: CastEfficiency,
    alwaysBeCasting: AlwaysBeCasting,
    cooldownTracker: CooldownTracker,

    // DoT uptimes
    agonyUptime: AgonyUptime,
    corruptionUptime: CorruptionUptime,
    siphonLifeUptime: SiphonLifeUptime,

    //Core
    soulShardTracker: SoulShardTracker,
    soulShardDetails: SoulShardDetails,
  };

  generateResults() {
    const results = super.generateResults();
    results.tabs = [
      {
        title: 'Suggestions',
        url: 'suggestions',
        render: () => (
          <SuggestionsTab issues={results.issues} />
        ),
      },
      {
        title: 'Talents',
        url: 'talents',
        render: () => (
          <Tab title="Talents">
            <Talents combatant={this.selectedCombatant} />
          </Tab>
        ),
      },
      // {
      //   title: 'Mana',
      //   url: 'mana',
      //   render: () => (
      //     <Tab title="Mana" style={{ padding: '15px 22px' }}>
      //       <Mana
      //         reportCode={this.report.code}
      //         actorId={this.playerId}
      //         start={this.fight.start_time}
      //         end={this.fight.end_time}
      //       />
      //     </Tab>
      //   ),
      // },
      ...results.tabs,
    ];

    return results;
  }
}

export default CombatLogParser;
