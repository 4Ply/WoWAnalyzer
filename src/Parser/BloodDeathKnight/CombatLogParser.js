import React from 'react';
import SuggestionsTab from 'Main/SuggestionsTab';
import Tab from 'Main/Tab';
import Talents from 'Main/Talents';
import CoreCombatLogParser from 'Parser/Core/CombatLogParser';

import CastEfficiency from './Modules/Features/CastEfficiency';
import AlwaysBeCasting from './Modules/Features/AlwaysBeCasting';
import CooldownTracker from './Modules/Features/CooldownTracker';
import BloodPlagueUptime from './Modules/Features/BloodPlagueUptime';
import BoneShieldUptime from './Modules/Features/BoneShieldUptime';
import OssuaryUptime from './Modules/Features/OssuaryUptime';
import WastedDeathAndDecay from './Modules/Features/WastedDeathAndDecay';
import RPWasted from './Modules/Features/RunicPowerWasted';
import BlooddrinkerTicks from './Modules/Features/BlooddrinkerTicks';

class CombatLogParser extends CoreCombatLogParser {

  static specModules = {

    //DeathKnight Core

    // Features
    castEfficiency: CastEfficiency,
    alwaysBeCasting: AlwaysBeCasting,
    cooldownTracker: CooldownTracker,
    BoneShieldUptime: BoneShieldUptime,
    OssuaryUptime: OssuaryUptime,
    WastedDeathAndDecay: WastedDeathAndDecay,
    RunicPowerWasted: RPWasted,
    BlooddrinkerTicks: BlooddrinkerTicks,

    //DOT
    bloodplagueUptime: BloodPlagueUptime,
    // Talents


    // Traits


    // Items:

  };

  generateResults() {
    const results = super.generateResults();

    results.tabs = [
      {
        title: 'Suggestions',
        url: 'suggestions',
        render: () => (
          <SuggestionsTab issues={results.issues} />
        ),
      },
      {
        title: 'Talents',
        url: 'talents',
        render: () => (
          <Tab title="Talents">
            <Talents combatant={this.modules.combatants.selected} />
          </Tab>
        ),
      },
      ...results.tabs,
    ];

    return results;
  }
}

export default CombatLogParser;
