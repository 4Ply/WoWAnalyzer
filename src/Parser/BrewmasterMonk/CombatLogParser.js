// TODO:

import React from 'react';

import MainCombatLogParser from 'Parser/Core/CombatLogParser';

import SuggestionsTab from 'Main/SuggestionsTab';
import Tab from 'Main/Tab';
import Talents from 'Main/Talents';

// Core
import HealingDone from './Modules/Core/HealingDone';
import DamageTaken from './Modules/Core/DamageTaken';
import DamageDone from './Modules/Core/DamageDone';
import HealingReceived from './Modules/Core/HealingReceived';
import Stagger from './Modules/Core/Stagger';
import IronSkinBrew from './Modules/Spells/IronSkinBrew';

// Features
import CastEfficiency from './Modules/Features/CastEfficiency';
import AlwaysBeCasting from './Modules/Features/AlwaysBeCasting';

class CombatLogParser extends MainCombatLogParser {
  static specModules = {
    // Core
    healingDone: HealingDone,
    healingReceived: HealingReceived,
    damageTaken: DamageTaken,
    stagger: Stagger,
    ironSkinBrew: IronSkinBrew,
    damageDone: DamageDone,
    
    // Features
    alwaysBeCasting: AlwaysBeCasting,
    castEfficiency: CastEfficiency,
    
    // Traits

    // Spells

    // Talents

    // Legendaries / Items:
  };

  generateResults() {
    const results = super.generateResults();

    results.tabs = [
      {
        title: 'Suggestions',
        url: 'suggestions',
        render: () => (
          <SuggestionsTab issues={results.issues} />
        ),
      },
      {
        title: 'Talents',
        url: 'talents',
        render: () => (
          <Tab title="Talents">
            <Talents combatant={this.selectedCombatant} />
          </Tab>
        ),
      },
      ...results.tabs,
    ];
    return results;
  }
}

export default CombatLogParser;
