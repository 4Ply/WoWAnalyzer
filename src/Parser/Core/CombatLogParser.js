import React from 'react';

import ITEMS from 'common/ITEMS';
import ItemLink from 'common/ItemLink';
import ItemIcon from 'common/ItemIcon';

import Combatants from './Modules/Combatants';
import AbilityTracker from './Modules/AbilityTracker';
import AlwaysBeCasting from './Modules/AlwaysBeCasting';
import Enemies from './Modules/Enemies';

// Shared Legendaries
import Prydaz from './Modules/Items/Prydaz';
import Velens from './Modules/Items/Velens';
import SephuzsSecret from './Modules/Items/SephuzsSecret';
// Shared Epics
import DrapeOfShame from './Modules/Items/DrapeOfShame';
import DarkmoonDeckPromises from './Modules/Items/DarkmoonDeckPromises';
import AmalgamsSeventhSpine from './Modules/Items/AmalgamsSeventhSpine';

import ParseResults from './ParseResults';
import SUGGESTION_IMPORTANCE from './ISSUE_IMPORTANCE';

function formatThousands(number) {
  return (Math.round(number || 0) + '').replace(/(\d)(?=(\d{3})+(?!\d))/g, "$1,");
}
function formatNumber(number) {
  if (number > 1000000) {
    return `${(number / 1000000).toFixed(2)}m`;
  }
  if (number > 10000) {
    return `${Math.round(number / 1000)}k`;
  }
  return formatThousands(number);
}
function formatPercentage(percentage) {
  return (Math.round((percentage || 0) * 10000) / 100).toFixed(2);
}
function getSuggestionImportance(value, regular, major, higherIsWorse = false) {
  if (higherIsWorse ? value > major : value < major) {
    return SUGGESTION_IMPORTANCE.MAJOR;
  }
  if (higherIsWorse ? value > regular : value < regular) {
    return SUGGESTION_IMPORTANCE.REGULAR;
  }
  return SUGGESTION_IMPORTANCE.MINOR;
}

class CombatLogParser {
  static abilitiesAffectedByHealingIncreases = [];

  static defaultModules = {
    combatants: Combatants,
    enemies: Enemies,
    abilityTracker: AbilityTracker,
    alwaysBeCasting: AlwaysBeCasting,

    // Items:
    // Legendaries:
    prydaz: Prydaz,
    velens: Velens,
    sephuzsSecret: SephuzsSecret,
    // Epics:
    drapeOfShame: DrapeOfShame,
    amalgamsSeventhSpine: AmalgamsSeventhSpine,
    darkmoonDeckPromises: DarkmoonDeckPromises,
  };
  // Override this with spec specific modules
  static specModules = {};

  report = null;
  player = null;
  fight = null;

  modules = {};

  get playerId() {
    return this.player.id;
  }

  /** @returns Combatants */
  get combatants() {
    return this.modules.combatants;
  }
  /** @returns {Combatant} */
  get selectedCombatant() {
    return this.combatants.selected;
  }

  get fightDuration() {
    return (this.finished ? this.fight.end_time : this.currentTimestamp) - this.fight.start_time;
  }

  _timestamp = null;
  get currentTimestamp() {
    return this._timestamp;
  }
  get playersById() {
    return this.report.friendlies.reduce((obj, player) => {
      obj[player.id] = player;
      return obj;
    }, {});
  }

  constructor(report, player, fight) {
    this.report = report;
    this.player = player;
    this.fight = fight;

    this.initializeModules(this.constructor.defaultModules);
    this.initializeModules(this.constructor.specModules);
  }
  initializeModules(modules) {
    Object.keys(modules).forEach(key => {
      const value = modules[key];
      // This may override existing modules, this is intended.
      this.modules[key] = new value(this);
    });
  }

  parseEvents(events) {
    return new Promise((resolve, reject) => {
      events.forEach(event => {
        if (this.error) {
          throw new Error(this.error);
        }
        this._timestamp = event.timestamp;

        // Triggering a lot of events here for development pleasure; does this have a significant performance impact?
        this.triggerEvent('event', event);
        this.triggerEvent(event.type, event);
        if (this.byPlayer(event)) {
          this.triggerEvent(`byPlayer_${event.type}`, event);
        }
        if (this.toPlayer(event)) {
          this.triggerEvent(`toPlayer_${event.type}`, event);
        }
      });

      resolve(events.length);
    });
  }
  triggerEvent(eventType, event) {
    const methodName = `on_${eventType}`;
    this.constructor.tryCall(this, methodName, event);
    Object.keys(this.modules)
      .map(key => this.modules[key])
      .sort((a, b) => b.priority - a.priority)
      .filter(module => module.active)
      .forEach(module => {
        this.constructor.tryCall(module, methodName, event);
      });
  }
  static tryCall(object, methodName, event) {
    const method = object[methodName];
    if (method) {
      method.call(object, event);
    }
  }

  byPlayer(event, playerId = this.player.id) {
    return (event.sourceID === playerId);
  }
  toPlayer(event, playerId = this.player.id) {
    return (event.targetID === playerId);
  }

  initialized = false;
  error = null;
  on_initialized() {
    this.initialized = true;
    if (!this.selectedCombatant) {
      this.error = 'The selected player could not be found in this fight. Make sure the log is recorded with Advanced Combat Logging enabled.';
    }
  }
  finished = false;
  on_finished() {
    this.finished = true;
  }

  // This used to be implemented as a sanity check, may be replaced by a cleaner solution.
  totalHealing = 0;
  on_byPlayer_heal(event) {
    this.totalHealing += event.amount + (event.absorbed || 0);
  }
  on_byPlayer_absorbed(event) {
    this.totalHealing += event.amount + (event.absorbed || 0);
  }
  totalDamage = 0;
  on_byPlayer_damage(event) {
    this.totalDamage += event.amount + (event.absorbed || 0);
  }
  // TODO: Damage taken from LOTM

  static SUGGESTION_VELENS_BREAKPOINT = 0.045;

  generateResults() {
    const results = new ParseResults();

    const fightDuration = this.fightDuration;
    const getPercentageOfTotal = healingDone => healingDone / this.totalHealing;
    const formatItemHealing = healingDone => `${formatPercentage(getPercentageOfTotal(healingDone))} % / ${formatNumber(healingDone / fightDuration * 1000)} HPS`;

    if (this.modules.prydaz.active) {
      results.items.push({
        id: ITEMS.PRYDAZ_XAVARICS_MAGNUM_OPUS.id,
        icon: <ItemIcon id={ITEMS.PRYDAZ_XAVARICS_MAGNUM_OPUS.id} />,
        title: <ItemLink id={ITEMS.PRYDAZ_XAVARICS_MAGNUM_OPUS.id} />,
        result: (
          <dfn data-tip="The effective healing contributed by the Prydaz, Xavaric's Magnum Opus equip effect.">
            {formatItemHealing(this.modules.prydaz.healing)}
          </dfn>
        ),
      });
    }
    if (this.modules.velens.active) {
      // TODO: Move this to the Velen's module? having it here stops making sense. If "item" modules could just provide their own item data object & suggestion then everything would be completely together. At that point the parser would have to recognize items automatically and add them to the items array as their modules are loaded. Doing this could be nice as items would be completely isolated in their module files, but I don't think mixing the currently pure JS modules with React is favorable.
      results.items.push({
        id: ITEMS.VELENS_FUTURE_SIGHT.id,
        icon: <ItemIcon id={ITEMS.VELENS_FUTURE_SIGHT.id} />,
        title: <ItemLink id={ITEMS.VELENS_FUTURE_SIGHT.id} />,
        result: (
          <dfn data-tip={`The effective healing contributed by the Velen's Future Sight use effect. ${formatPercentage(this.modules.velens.healingIncreaseHealing/this.totalHealing)}% of total healing was contributed by the 15% healing increase and ${formatPercentage(this.modules.velens.overhealHealing/this.totalHealing)}% of total healing was contributed by the overhealing distribution.`}>
            {formatItemHealing(this.modules.velens.healing)}
          </dfn>
        ),
      });
      const velensHealingPercentage = getPercentageOfTotal(this.modules.velens.healing);
      if (velensHealingPercentage < this.constructor.SUGGESTION_VELENS_BREAKPOINT) {
        results.addIssue({
          issue: <span>Your usage of <ItemLink id={ITEMS.VELENS_FUTURE_SIGHT.id} /> can be improved. Try to maximize the amount of healing during the buff without excessively overhealing on purpose, or consider using an easier legendary ({(velensHealingPercentage * 100).toFixed(2)}% healing contributed).</span>,
          icon: ITEMS.VELENS_FUTURE_SIGHT.icon,
          importance: getSuggestionImportance(velensHealingPercentage, this.constructor.SUGGESTION_VELENS_BREAKPOINT - 0.005, this.constructor.SUGGESTION_VELENS_BREAKPOINT - 0.015),
        });
      }
    }
    if (this.modules.drapeOfShame.active) {
      results.items.push({
        id: ITEMS.DRAPE_OF_SHAME.id,
        icon: <ItemIcon id={ITEMS.DRAPE_OF_SHAME.id} />,
        title: <ItemLink id={ITEMS.DRAPE_OF_SHAME.id} />,
        result: (
          <dfn data-tip="The effective healing contributed by the critical healing gain of the Drape of Shame equip effect.">
            {formatItemHealing(this.modules.drapeOfShame.healing)}
          </dfn>
        ),
      });
    }
    if (this.modules.sephuzsSecret.active) {
      results.items.push({
        id: ITEMS.SEPHUZS_SECRET.id,
        icon: <ItemIcon id={ITEMS.SEPHUZS_SECRET.id} />,
        title: <ItemLink id={ITEMS.SEPHUZS_SECRET.id} />,
        result: `${((this.modules.sephuzsSecret.uptime / fightDuration * 100) || 0).toFixed(2)} % uptime`,
      });
    }
    if (this.modules.amalgamsSeventhSpine.active) {
      results.items.push({
        id: ITEMS.AMALGAMS_SEVENTH_SPINE.id,
        icon: <ItemIcon id={ITEMS.AMALGAMS_SEVENTH_SPINE.id} />,
        title: <ItemLink id={ITEMS.AMALGAMS_SEVENTH_SPINE.id} />,
        result: (
          <dfn data-tip={`The exact amount of mana gained from the Amalgam's Seventh Spine equip effect. The buff expired successfully ${this.modules.amalgamsSeventhSpine.procs} times and the buff was refreshed ${this.modules.amalgamsSeventhSpine.refreshes} times (refreshing delays the buff expiration and is inefficient use of this trinket).`}>
            {formatThousands(this.modules.amalgamsSeventhSpine.manaGained)} mana gained ({formatThousands(this.modules.amalgamsSeventhSpine.manaGained / this.fightDuration * 1000 * 5)} MP5)
          </dfn>
        ),
      });
    }
    if (this.modules.darkmoonDeckPromises.active) {
      results.items.push({
        id: ITEMS.DARKMOON_DECK_PROMISES.id,
        icon: <ItemIcon id={ITEMS.DARKMOON_DECK_PROMISES.id} />,
        title: <ItemLink id={ITEMS.DARKMOON_DECK_PROMISES.id} />,
        result: (
          <dfn data-tip="The exact amount of mana saved by the Darkmoon Deck: Promises equip effect. This takes the different values per card into account at the time of the cast.">
            {formatThousands(this.modules.darkmoonDeckPromises.manaGained)} mana saved ({formatThousands(this.modules.darkmoonDeckPromises.manaGained / this.fightDuration * 1000 * 5)} MP5)
          </dfn>
        ),
      });
    }

    return results;
  }
}

export default CombatLogParser;
