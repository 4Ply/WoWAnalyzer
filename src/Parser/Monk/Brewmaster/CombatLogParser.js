import CoreCombatLogParser from 'Parser/Core/CombatLogParser';
import DamageDone from 'Parser/Core/Modules/DamageDone';
// Core
import HealingDone from './Modules/Core/HealingDone';
import DamageTaken from './Modules/Core/DamageTaken';
import HealingReceived from './Modules/Core/HealingReceived';
import Stagger from './Modules/Core/Stagger';
import StaggerFabricator from './Modules/Core/StaggerFabricator';
// Spells
import IronSkinBrew from './Modules/Spells/IronSkinBrew';
import BlackoutCombo from './Modules/Spells/BlackoutCombo';
// Features
import Checklist from './Modules/Features/Checklist';
import Abilities from './Modules/Features/Abilities';
import AlwaysBeCasting from './Modules/Features/AlwaysBeCasting';
// Items
import T20_2pc from './Modules/Items/T20_2pc';
import T20_4pc from './Modules/Items/T20_4pc';
// normalizers
import IronskinBrewNormalizer from './Modules/Normalizers/IronskinBrew';

class CombatLogParser extends CoreCombatLogParser {
  static specModules = {
    // Core
    healingDone: [HealingDone, { showStatistic: true }],
    healingReceived: HealingReceived,
    damageTaken: [DamageTaken, { showStatistic: true }],
    stagger: Stagger,
    staggerFabricator: StaggerFabricator,
    damageDone: [DamageDone, { showStatistic: true }],

    // Features
    checklist: Checklist,
    alwaysBeCasting: AlwaysBeCasting,
    abilities: Abilities,

    // Spells
    ironSkinBrew: IronSkinBrew,
    blackoutCombo: BlackoutCombo,

    // Items
    t20_2pc: T20_2pc,
    t20_4pc: T20_4pc,

    // normalizers
    isbNormalizer: IronskinBrewNormalizer,
  };
}

export default CombatLogParser;
