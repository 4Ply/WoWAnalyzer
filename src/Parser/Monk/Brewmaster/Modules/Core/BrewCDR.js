import React from 'react';
import SPELLS from 'common/SPELLS';
import SpellIcon from 'common/SpellIcon';
import { formatPercentage } from 'common/format';
import Combatants from 'Parser/Core/Modules/Combatants';
import Analyzer from 'Parser/Core/Analyzer';
import StatisticBox, { STATISTIC_ORDER } from 'Main/StatisticBox';
import KegSmash from '../Spells/KegSmash';
import TigerPalm from '../Spells/TigerPalm';
import BlackOxBrew from '../Spells/BlackOxBrew';
import AnvilHardenedWristwraps from '../Items/AnvilHardenedWristwraps';

class BrewCDR extends Analyzer {
  static dependencies = {
    combatants: Combatants,
    ks: KegSmash,
    tp: TigerPalm,
    wrists: AnvilHardenedWristwraps,
    bob: BlackOxBrew,
  }

  get totalCDR() {
    let totalCDR = 0;
    // add in KS CDR...
    totalCDR += this.ks.cdr;
    totalCDR += this.ks.bocCDR;
    // ...and TP...
    totalCDR += this.tp.cdr + this.tp.fpCDR;
    // ...and BoB...
    totalCDR += this.bob.cdr;
    // ...and wrists
    totalCDR += this.wrists.cdr;
    return totalCDR;
  }

  get cooldownReductionRatio() {
    return 1.0 - this.owner.fightDuration / (this.owner.fightDuration + this.totalCDR);
  }

  get suggestionThreshold() {
    return {
      actual: this.cooldownReductionRatio,
      isLessThan: {
        minor: 0.5,
        average: 0.45,
        major: 0.4,
      },
      style: 'percentage',
    };
  }

  statistic() {
    let wristsDesc = "";
    if(this.wrists.active) {
      wristsDesc = `<li>Anvil-Hardened Wristwraps and ${this.wrists.dodgedHits} dodged hits — <b>${(this.wrists.cdr / 1000).toFixed(2)}s</b> (<b>${(this.wrists.wastedCDR / 1000).toFixed(2)}s</b> wasted)</li>`;
    }
    let bobDesc = "";
    if(this.bob.active) {
      bobDesc = `<li>${this.bob.casts} Black Ox Brew casts -- <b>${(this.bob.cdr / 1000).toFixed(2)}s</b> (<b>${(this.bob.wastedCDR / 1000).toFixed(2)}s</b> wasted)</li>`;
    }
    let bocKsDesc = "";
    if(this.ks.bocHits > 0) {
      bocKsDesc = `<li>Using Blackout Combo on ${this.ks.bocHits} Keg Smash hits — <b>${(this.ks.bocCDR / 1000).toFixed(2)}s</b> (<b>${(this.ks.wastedBocCDR / 1000).toFixed(2)}s</b> wasted)</li>`;
    }
    return (
      <StatisticBox icon={<SpellIcon id={SPELLS.TIGER_PALM.id} />}
        value={`${formatPercentage(this.cooldownReductionRatio)}%`}
        label="Effective Brew CDR"
        tooltip={`Your cooldowns were reduced by: <ul>
              <li>${this.ks.totalHits} Keg Smash hits (${(this.ks.totalHits / this.ks.totalCasts).toFixed(2)} per cast) — <b>${(this.ks.cdr / 1000).toFixed(2)}s</b> (<b>${(this.ks.wastedCDR / 1000).toFixed(2)}s</b> wasted)</li>
              ${bocKsDesc}
              <li>${this.tp.totalCasts} Tiger Palm hits (with ${this.tp.facePalmHits} Face Palm procs) — <b>${((this.tp.cdr + this.tp.fpCDR) / 1000).toFixed(2)}s</b> (<b>${((this.tp.wastedCDR + this.tp.wastedFpCDR) / 1000).toFixed(2)}s</b> wasted)</li>
              ${bobDesc}
              ${wristsDesc}
            </ul>
            <b>Total cooldown reduction:</b> ${(this.totalCDR / 1000).toFixed(2)}s.</b>`}
      />
    );
  }
  statisticOrder = STATISTIC_ORDER.OPTIONAL();
}

export default BrewCDR;
