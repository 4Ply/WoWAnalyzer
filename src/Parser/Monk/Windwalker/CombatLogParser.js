import CoreCombatLogParser from 'Parser/Core/CombatLogParser';
// Features
import DamageDone from 'Parser/Core/Modules/DamageDone';
import AlwaysBeCasting from './Modules/Features/AlwaysBeCasting';
import Abilities from './Modules/Features/Abilities';
import CooldownThroughputTracker from './Modules/Features/CooldownThroughputTracker';
import Checklist from './Modules/Features/Checklist';
// Chi
import ChiDetails from './Modules/Chi/ChiDetails';
import ChiTracker from './Modules/Chi/ChiTracker';
// Core
import Channeling from './Modules/Core/Channeling';
// Spells
import ComboBreaker from './Modules/Spells/ComboBreaker';
import StormEarthAndFire from './Modules/Spells/StormEarthAndFire';
import FistsofFury from './Modules/Spells/FistsofFury';
import SpinningCraneKick from './Modules/Spells/SpinningCraneKick';
import ComboStrikes from './Modules/Spells/ComboStrikes';
import TouchOfKarma from './Modules/Spells/TouchOfKarma';
import BlackoutKick from './Modules/Spells/BlackoutKick';
// Talents
import HitCombo from './Modules/Talents/HitCombo';
import EnergizingElixir from './Modules/Talents/EnergizingElixir';

class CombatLogParser extends CoreCombatLogParser {
  static specModules = {
    // Core
    channeling: Channeling,

    // Features
    damageDone: [DamageDone, { showStatistic: true }],
    alwaysBeCasting: AlwaysBeCasting,
    abilities: Abilities,
    cooldownThroughputTracker: CooldownThroughputTracker,
    checklist: Checklist,

    // Resources
    chiTracker: ChiTracker,
    chiDetails: ChiDetails,

    // Talents:
    hitCombo: HitCombo,
    energizingElixir: EnergizingElixir,

    // Spells;
    comboBreaker: ComboBreaker,
    stormEarthAndFire: StormEarthAndFire,
    fistsofFury: FistsofFury,
    spinningCraneKick: SpinningCraneKick,
    touchOfKarma: TouchOfKarma,
    comboStrikes: ComboStrikes,
    blackoutKick: BlackoutKick,
  };
}

export default CombatLogParser;
