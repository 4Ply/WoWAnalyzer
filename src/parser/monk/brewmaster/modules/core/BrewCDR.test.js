import snapshotTest from 'parser/core/tests/snapshotTest';
import CombatLogParser from '../../CombatLogParser';
import BrewCDR from './BrewCDR';

describe('BrewCDR', () => {
  snapshotTest(CombatLogParser, BrewCDR, 'brm-example');
  snapshotTest(CombatLogParser, BrewCDR, 'brm-example', function cooldownReductionRatio(ana) { return ana.cooldownReductionRatio; });
});
