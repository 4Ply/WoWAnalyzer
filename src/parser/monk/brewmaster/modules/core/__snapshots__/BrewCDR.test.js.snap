// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BrewCDR should match the cdr snapshot 1`] = `0.34592463516500815`;

exports[`BrewCDR should match the statistic snapshot 1`] = `
<div
  className="col-lg-3 col-md-4 col-sm-6 col-xs-12"
  style={
    Object {
      "zIndex": 1,
    }
  }
>
  <div
    className="panel statistic-box expandable"
  >
    <div
      className="panel-body"
    >
      <div
        className="flex"
      >
        <div
          className="flex-sub statistic-icon"
          style={
            Object {
              "alignItems": "center",
              "display": "flex",
            }
          }
        >
          <a
            href="http://wowhead.com/spell=100780"
            rel="noopener noreferrer"
            target="_blank"
          >
             
            <img
              alt=""
              className="icon"
              src="//render-us.worldofwarcraft.com/icons/56/ability_monk_tigerpalm.jpg"
            />
          </a>
        </div>
        <div
          className="flex-main"
          style={
            Object {
              "paddingLeft": 16,
              "position": "relative",
            }
          }
        >
          <div
            className="slabel"
          >
            Effective Brew CDR
          </div>
          <div
            className="value"
          >
            <dfn
              data-tip="Your cooldowns were reduced by: <ul>
              <li>31 Keg Smash casts— <b>105.59s</b> (<b>18.41s</b> wasted)</li>
              
              <li>66 Tiger Palm hits — <b>57.00s</b> (<b>9.00s</b> wasted)</li>
              
            </ul>
            <b>Total cooldown reduction:</b> 162.59s.</b><br/>
            <b>Minimum Cooldown Reduction for 100% ISB uptime:</b> 25.88%"
            >
              34.59%
            </dfn>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
